src/Core_Fix_StudioOptimizations/MeasuringSceneLoadTimes.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using BepInEx.Logging;
+using HarmonyLib;
+using Studio;
+
+namespace IllusionFixes
+{
+    /// <summary>
+    /// Outputs scene load times to log
+    /// </summary>
+    class MeasuringLoadTimes
+    {
+        private static ManualLogSource Logger;
+
+        private static DateTime _startLoading = new DateTime();
+        private static bool _loading = false;
+
+        public static void Setup( ManualLogSource logger )
+        {
+            Logger = logger;
+            UnityEngine.SceneManagement.SceneManager.sceneUnloaded += OnSceneUnloaded;
+            Harmony.CreateAndPatchAll(typeof(MeasuringLoadTimes), nameof(MeasuringLoadTimes));
+        }
+
+        [HarmonyPrefix]
+        [HarmonyPriority(Priority.High)]
+        [HarmonyPatch(typeof(SceneLoadScene), "OnClickLoad")]
+        private static void OnClickLoadPrefix()
+        {
+            OnStartLoadScene();
+        }
+
+        [HarmonyPrefix]
+        [HarmonyPriority(Priority.High)]
+        [HarmonyPatch(typeof(SceneLoadScene), "OnClickImport")]
+        private static void OnClickImportPrefix()
+        {
+            OnStartLoadScene();
+        }
+
+        static private void OnSceneUnloaded(UnityEngine.SceneManagement.Scene arg0)
+        {
+            if (_loading && arg0.name == "StudioSceneLoad")
+            {
+                _loading = false;
+                DateTime end = DateTime.Now;
+                double sec = (end - _startLoading).TotalSeconds;
+                Logger.LogInfo($"Scene Loaded: {sec:F2}[s]");
+            }
+        }
+
+        static private void OnStartLoadScene()
+        {
+            _startLoading = System.DateTime.Now;
+            _loading = true;
+        }
+    }
+}

```

Comment:
Measuring time can be done much more simply by passing a StopWatch instance between prefix and postfix patches like here: https://github.com/IllusionMods/IllusionFixes/blob/d60f4e37ce864980bc371aae962c89ddc7801abf/src/KKS_Fix_CharacterListOptimizations/CharacterListOptimizations.cs#L44-L56
It would also make it no longer necessary to have completely separate class for this.

Is the rest of the PR ready for a review?

Reply:
System.Diagnostics.StopWatch is not in KK?
Is StopWatch in some namespace?

Oh sorry. No more changes. Please review.

Reply:
> It would also make it no longer necessary to have completely separate class for this.

This was intentionally separated.
I knew that if I kept adding to StudioOptimizations I would end up with spaghetti one day.
Wouldn't it be better to separate the files by appropriate size?

Reply:
I think splitting the patches out into partial classes is fine, that way there is still only one Harmony.PatchAll call which is the most expensive thing during initialization.

The timing could be done like this instead (not tested but should work):
```suggestion
    public partial class StudioOptimizations
    {
        [HarmonyPrefix]
        [HarmonyPriority(Priority.First)]
        [HarmonyPatch(typeof(Studio), nameof(Studio.LoadScene))]
        [HarmonyPatch(typeof(Studio), nameof(Studio.ImportScene))]
        private static void LoadScenePrefix(out Stopwatch __state)
        {
            __state = Stopwatch.StartNew(); 
        }

        [HarmonyPostfix]
        [HarmonyPriority(Priority.Last)]
        [HarmonyPatch(typeof(Studio), nameof(Studio.LoadScene))]
        [HarmonyPatch(typeof(Studio), nameof(Studio.ImportScene))]
        private static void LoadScenePostfix(Stopwatch __state)
        {
            Logger.LogDebug("Scene took " + __state.ElapsedMilliseconds + "ms to load/import"); 
        }
        
        // Coroutine needs a different approach
        [HarmonyPostfix]
        [HarmonyPriority(Priority.Last)]
        [HarmonyPatch(typeof(Studio), nameof(Studio.LoadSceneCoroutine))]
        private static void LoadSceneCoroutineWrapper(ref IEnumerator __result)
        {
            IEnumerator WrapperCo (IEnumerator original)
            {
                var sw = Stopwatch.StartNew();
                yield return original;
                Logger.LogDebug("Scene took " + sw.ElapsedMilliseconds + "ms to load"); 
            }
            
            __result = WrapperCo(__result);
        }
        
    }
}

```

Reply:
I understand about the class splitting, I will change it to use Partial.
Personally, though, I would prefer class splitting to the cost of PatchAll.

I didn't know there was such a thing. Thanks.
No, but does this include IsFileValid()?
IsFileValid() is heavy and I'd like to include it in the measurement.

I'll give it a try.

Reply:
LoadScene/ImportScene prefix/postfix did not work in KKS.
It seems that LoadScene/ImportScene is not being called.

The reason why the build of Stopwatch did not pass in KK was because there were not enough references to System.
I added System to KK and modified it to use Stopwatch.

Wrapping LoadSceneCoroutine() had the following problems
- Not including IsFileValid
- The call is a few seconds earlier than OnSceneUnloaded; it seems that some processing is done after the end of LoadSceneCoroutine().

IsFileValid may not be loading the scene correctly, but I think it should be included since it is associated with it.

Reply:
The issue with OnClickLoad and OnClickImport is that it doesn't handle scenes loaded by other plugins, like the custom scene browser plugin.
 
LoadScene is not called by the base game anywhere in KKS, LoadSceneCoroutine is used instead. Plugins can still use the first one since it works fine so it should be checked as well.

You're right, IsFileValid is checked only on the UI buttons. I forgot why it was done this way but it's not ideal, it should happen at the start of LoadScene and such instead, but I think there was some sort of a problem preventing it.

Reply:
Hmm?
Should I give up on IsFileValid and split it up as follows?

KKS: LoadSceneCoroutine wrapper
Non-KKS: LoadScene/ImportScene patch

Reply:
I think the current version is good. If there's a need in the future it can be changed easily.